### PR TITLE
os/bluestore: keep statfs replica in RAM to avoid expensive KV retrieval

### PR DESCRIPTION
**Pull request by ifed01**
Introduce some debug switches as well.
Signed-off-by: Igor Fedotov <ifedotov@mirantis.com>

### CODE REVIEW

**Comment by liewegas** [comment]
retest this please

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -3973,7 +3973,7 @@ int BlueStore::_open_fm(bool create)
 	start += l + u;
       }
     }
-    db->submit_transaction_sync(t);
+    cct->_conf->bluestore_debug_omit_kv_commit ? 0 : db->submit_transaction_sync(t);
```
Comment (varadakari) [inline src/os/bluestore/BlueStore.cc]: may be we can let the submit_tx() get through, this part of the code is mostly on boot up time. Allocators and free list manager might need to store the info in DB.  If we want to measure any performance changes after bootup the debug flag will help. Similar in mkfs().

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -7445,6 +7454,11 @@ void BlueStore::_txc_update_store_statfs(TransContext *txc)
   logger->inc(l_bluestore_compressed_allocated, txc->statfs_delta.compressed_allocated());
   logger->inc(l_bluestore_compressed_original, txc->statfs_delta.compressed_original());
 
+  {
+    std::lock_guard<std::mutex> l(vstatfs_lock);
```
Comment (varadakari) [inline src/os/bluestore/BlueStore.cc]: @ifed01 did you do benchmark runs with this lock introduction? Mostly will be better than reading from DB, did we observe any improvement?

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -5002,6 +5024,8 @@ int BlueStore::_mount(bool kv_only)
   if (r < 0)
     goto out_alloc;
 
+  open_statfs();
+
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: need the same thing in fsck().  Or better yet just put it in _open_super_meta().

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: fixed

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: fixed

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: @varadakari - No, I don't expect any performance difference here - single lock acquisition is a pretty simple operation, we have tons of them along the write path. This was mainly to resolve Sage's complaint on significant amount of time spent in BlueStore::statfs func

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-31_18:45:30-rados-wip-sage-testing---basic-smithi/

**Comment by liewegas** [comment]
rebase please?

**Comment by ifed01** [comment]
rebased!

**Comment by ifed01** [comment]
retest this please

**Comment by ifed01** [comment]
@liewegas - ping

**Comment by liewegas** [comment]
Thanks, I lost track of this!

**Pull request merged**